drivers/potentiometer/ad5293/ad5293.h

```diff
@@ -0,0 +1,149 @@
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef _ad5293_H_
+#define _ad5293_H_
+/*****************************************************************************/
+/****************************** Include Files ********************************/
+/*****************************************************************************/
+#include <stdint.h>
+#include "no_os_gpio.h"
+#include "no_os_spi.h"
+#include "no_os_util.h"
+
+/* SPI Read/Write commands */
+#define CMD_NOP        0b0000
+#define CMD_W_WIPER    0b0001
+#define CMD_R_WIPER    0b0010
+#define CMD_RESET      0b0100
+#define CMD_W_CTRL     0b0110
+#define CMD_R_CTRL     0b0111
+#define CMD_PD         0b1000
+#define CMD_SDO_Z      0x8001
+
+/* Bit mask*/
+#define CALI_MSK       NO_OS_BIT(2)
+#define PROTECT_MSK    NO_OS_BIT(1)
+
+/* calibration mode */
+enum calibration_mode_t {
+	CALI_PERFORMANCE,
+	CALI_NORMAL
+};
+
+/* protect mode */
+enum protect_mode_t {
+	PROTECT_LOCK,
+	PROTECT_UNLOCK // enable update wiper position through digital interface
+};
+
+/* shutdown mode */
+enum shutdown_t {
+	NORMAL,
+	SHUTDOWN
+};
+
+/* operation mode */
+enum operation_mode_t {
+	BUS,
+	DAISY_CHAIN
+};
+
+/* SDO mode */
+enum SDO_mode_t {
+	SDO_DISENABLE,
+	SDO_ENABLE
+};
+
+/* SPI frame */
+#define AD5293_W_CTRL_FRAME(pm,cali)     (uint16_t)((CMD_W_CTRL<<10)|(pm<<1)|(cali<<2))
+#define AD5293_R_CTRL_FRAME              (uint16_t)CMD_R_CTRL<<10
+#define AD5293_W_WIPER_FRAME(data)       (uint16_t)((CMD_W_WIPER<<10)|data)
+#define AD5293_R_WIPER_FRAME             (uint16_t)CMD_R_WIPER<<10
+#define SHUTDOWN_FRAME(sd)               (uint16_t)((CMD_PD<<10)|sd)
+#define RESET_FRAME                      (uint16_t)CMD_RESET<<10
+
+/**********************************struct*****************************************/
+struct ad5293_chip_info {
+	enum calibration_mode_t cali;
+	uint16_t Wiper_value; //10bit
+};
+
+struct ad5293_dev {
+	/* SPI */
+	struct no_os_spi_desc	*spi_desc;
+	/* GPIO */
+	struct no_os_gpio_desc	*gpio_reset;
+	/* number of chips */
+	uint16_t chip_num;
+	/* pointer of chip struct  */
+	struct ad5293_chip_info* chip; //point to chip 0
+};
+
+struct ad5293_init_param {
+	/* SPI */
+	struct no_os_spi_init_param spi_init;
+	/* GPIO */
+	struct no_os_gpio_init_param* gpio_reset;
+	/* number of chips */
```

Comment:
same as above, is this the number of chips in daisy chain config ? 

i'd prefer it if the default uninitialized value of 0 would imply 1 chip

Reply:
I would say value 0 imply 1 is fine for array, but 0 imply for 1 chip is kinda anti-intuitive. I would chose do a better documentation that explains this well. Do you insist on that?
Daisy chain is only supported for multiple chips, because I cant figure a clearly/comprehensive way to switch CS line configuration on spi_desc( especially when port changed).
In the way I wrote single and daisy chain mode are compatible, only difference is several more bytes needs to be set for SPI tranfer buffer, I think this is better.

Reply:
My reasoning is one may use this device driver with 1-chip only, no daisy chain configuration needed, so they shouldn't bother to populate the ad5293_init_param.chip_num. So chip_num could be interpreted at 2 levels:
1. `if (chip_num)`, easy test that you will likely do in a few places in the device driver
2. chip_num value should mean the number of devices in daisy chain (document this as you proposed!): 0 - don't bother because of the test, 1 - you could treat it as invalid in the init() as it doesn't make sense, >=2 - use them to compute how many bytes need to be transferred

Reply:
Oh, I guess we have a misunderstanding here, since you are worrying 1 chip use. Let me explain clearer, 
1. if chip num is 1, hardware schamatic is normal SPI device topology. In init function 1 chip info struct space is allocated  for chip info. In drivers functions, sz indicate size of tranfer buffer size, so 2 is set. For communications 2 bytes will be shifted out.

memory allocate
>     chp = (struct ad5293_chip_info*)no_os_malloc(dev->chip_num * sizeof(* chp));
buffer array declare and size compute
> 	uint16_t sz = 2 * dev->chip_num;
> 	uint8_t buf[sz];
> 
> 	for (i = 0; i < dev->chip_num; i++) {
> 		buf[2 * i] = (uint8_t)(AD5293_R_WIPER_FRAME >> 8);
> 		buf[2 * i + 1] = (uint8_t)(AD5293_R_WIPER_FRAME);
> 	}
2. if chip num greater than 1,  hardware schamatic is required to be daisy chain. In init function `num` chip info space will be allocated for multiple chip. Driver functions will calculate proper size for data buffer, which is `2*num`, and there is no need to change anything because `no_os_spi_write_and_read` api keep CS line low until all `2*num` bytes are shifted.

The thing needs to change is the `chip_num` parameter for init struct and proper hardware topology, all programs work for both 1 chip or multiple chips set without necessary testing in operation functions. This is what meant by compatible. To avoid confusing I will add more description on README.